### PR TITLE
[WIP] [SPARK-35572] [K8S] add hostNetwork feature to executor

### PR DESCRIPTION
**Pull request by zwangsheng**
### What changes were proposed in this pull request?
add hostNetwork feature to executor
modify BasicExecutorFeatureStep add hostNetwork in executorpod 

### Why are the changes needed?
In the process of the company's business promotion, the function of k8s hostNetwork is used. But it is found that spark does not support enable hostNetwork in the configuration.
Found that this function will still be commonly used, so I decided to add a perceptible parameter to enable hostNetwork in spark way.

### Does this PR introduce _any_ user-facing change?
yes, add new config spark.kubernetes.executor.hostNetwork.enable

### How was this patch tested?
add unit test

### CODE REVIEW

**Comment by zwangsheng** [comment]
cc @yaooqinn

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 1]
```diff
@@ -473,6 +473,13 @@ private[spark] object Config extends Logging {
   val KUBERNETES_AUTH_SUBMISSION_CONF_PREFIX =
     "spark.kubernetes.authenticate.submission"
 
+  val KUBERNETES_EXECUTOR_HOSTNETWORK =
+    ConfigBuilder("spark.kubernetes.executor.hostNetwork.enable")
+    .doc("executor pod with hostNetwork status")
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:478]: indentation?

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala [hunk 2]
```diff
@@ -370,6 +370,15 @@ class BasicExecutorFeatureStepSuite extends SparkFunSuite with BeforeAndAfter {
     }
   }
 
+  test("test enable hostNetwork") {
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala:373]: Please add a JIRA ID prefix.
```
- test("test enable hostNetwork") {
+ test("SPARK-35572: test enable hostNetwork") {
```

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh [hunk 3]
```diff
@@ -67,6 +67,14 @@ elif ! [ -z ${SPARK_HOME+x} ]; then
   SPARK_CLASSPATH="$SPARK_HOME/conf:$SPARK_CLASSPATH";
 fi
 
+# when user want to formulate the IP they need, they can use this conf to post host
+if ! [ -z "${SPARK_EXECUTOR_POD_IP_OVERWRITE}" ]; then
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh:71]: What is this `SPARK_EXECUTOR_POD_IP_OVERWRITE`?

Comment (zwangsheng) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh:71]: it is a executor pod env, maybe we should describe it alone like ENV_DRIVER_URL？

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala [hunk 4]
```diff
@@ -45,10 +45,15 @@ private[spark] class BasicExecutorFeatureStep(
   private val blockManagerPort = kubernetesConf
     .sparkConf
     .getInt(BLOCK_MANAGER_PORT.key, DEFAULT_BLOCKMANAGER_PORT)
+  private val hostNetwork = kubernetesConf.get(KUBERNETES_EXECUTOR_HOSTNETWORK)
```
Comment (yaooqinn) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:48]: `hostNetwork` might be enabled when this is true or executor pod template contains right

**Comment by zwangsheng** [comment]
this SPARK_EXECUTOR_POD_IP_OVERWRITE has nothing to do with this PR, it will be removed in this PR first

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala [hunk 5]
```diff
@@ -370,7 +370,7 @@ class BasicExecutorFeatureStepSuite extends SparkFunSuite with BeforeAndAfter {
     }
   }
 
-  test("test enable hostNetwork") {
+  test("SPARK-35572 test enable hostNetwork") {
```
Comment (yaooqinn) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala:373]: `SPARK-35572: Configure executor pods with host network`

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala [hunk 6]
```diff
@@ -370,6 +370,15 @@ class BasicExecutorFeatureStepSuite extends SparkFunSuite with BeforeAndAfter {
     }
   }
 
+  test("SPARK-35572 test enable hostNetwork") {
+    baseConf.set(KUBERNETES_EXECUTOR_HOSTNETWORK, true)
+    baseConf.set(BLOCK_MANAGER_PORT, 0)
+    val baseDriverPod = SparkPod.initialPod()
+    val basicExecutorPod = new BasicExecutorFeatureStep(
+      newExecutorConf(), new SecurityManager(baseConf), defaultProfile).configurePod(baseDriverPod)
+    assert(basicExecutorPod.pod.getSpec.getHostNetwork == true)
```
Comment (yaooqinn) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala:379]: `assert(basicExecutorPod.pod.getSpec.getHostNetwork)`

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala [hunk 7]
```diff
@@ -45,10 +45,15 @@ private[spark] class BasicExecutorFeatureStep(
...
 
   require(blockManagerPort == 0 || (1024 <= blockManagerPort && blockManagerPort < 65536),
     "port number must be 0 or in [1024, 65535]")
 
+  if (hostNetwork) {
+    require(blockManagerPort == 0, "when enable hostNetwork executor port should be random")
```
Comment (yaooqinn) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:54]: we  also need to cover this line in test

**Comment by yaooqinn** [comment]
ok to test

**Comment by zwangsheng** [comment]
cc @yaooqinn

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile [hunk 8]
```diff
@@ -31,7 +31,7 @@ RUN set -ex && \
     sed -i 's/http:\/\/deb.\(.*\)/https:\/\/deb.\1/g' /etc/apt/sources.list && \
     apt-get update && \
     ln -s /lib /lib64 && \
-    apt install -y bash tini libc6 libpam-modules krb5-user libnss3 procps && \
+    apt install -y bash tini libc6 libpam-modules krb5-user libnss3 procps dnsutils && \
```
Comment (yaooqinn) [inline resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile:34]: What is `dnsutils` for?

Comment (zwangsheng) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: will remove this

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala [hunk 9]
```diff
@@ -370,6 +370,20 @@ class BasicExecutorFeatureStepSuite extends SparkFunSuite with BeforeAndAfter {
     }
   }
 
+  test("SPARK-35572: Configure executor pods with host network enable") {
+    baseConf.set(KUBERNETES_EXECUTOR_HOSTNETWORK, true)
+    val baseDriverPod = SparkPod.initialPod()
```
Comment (yaooqinn) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala:375]: baseDriverPod？

Comment (zwangsheng) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala]: fixed

**Comment by holdenk** [comment]
I'm a little concerned about adding this given all of the security considerations around hostNetwork on Kube. If we do add this I think we should make it very very clear that turning this on has negative security implications.

**Pull request closed without merging**